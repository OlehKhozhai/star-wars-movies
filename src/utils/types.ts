export type Movie = {
  title: string;
  episode_id: number;
  opening_crawl: string;
  director: string;
  producer: string;
  release_date: string;
  created: string;
  url: string;
};

export type FavoriteMovies = Record<string, boolean>;
