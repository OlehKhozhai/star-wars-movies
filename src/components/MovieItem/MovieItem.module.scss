.root {
  cursor: pointer;
  position: relative;
  padding: 15px 40px;
}

.active {
  background-color: #e1e1e1;
  border-radius: 5px;
}

.starIcon {
  position: absolute;
  top: 10px;
  right: 10px;
}
