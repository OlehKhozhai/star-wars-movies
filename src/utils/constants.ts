export enum APIPath {
  Films = "https://swapi.dev/api/films",
}


export enum LSKey {
  FavoriteMovies = "FAVORITE_MOVIES",
}