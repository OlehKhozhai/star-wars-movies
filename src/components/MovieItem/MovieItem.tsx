import { memo, FC } from "react";

import { StarIcon } from "assets/icons/StarIcon";
import { Movie } from "utils/types";
import styles from "./MovieItem.module.scss";

type MovieItemProps = Movie & {
  isActive: boolean;
  isFavorite: boolean;
  onMovieClick: (movie: Movie) => void;
};

const MovieItem: FC<MovieItemProps> = ({
  isActive,
  isFavorite,
  onMovieClick,
  ...movie
}) => {
  const handleMovieClick = () => {
    onMovieClick(movie);
  };

  return (
    <li
      className={`${styles.root} ${isActive ? styles.active : ""}`}
      onClick={handleMovieClick}
    >
      {movie.title}

      <span className={styles.starIcon}>
        <StarIcon fill={isFavorite ? "yellow" : "grey"} />
      </span>
    </li>
  );
};

export default memo(MovieItem);
