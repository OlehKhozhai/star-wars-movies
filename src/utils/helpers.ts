import { LSKey } from "utils/constants";
import { FavoriteMovies, Movie } from "utils/types";

export const getAPIData = <T>(
  url: string,
  params: RequestInit = {}
): Promise<T> =>
  fetch(url, { ...params })
    .then((data) => data.json())
    .then((data) => data)
    .catch((error) => console.log("--error--", error));

export const initializeFavoriteMoviesConfigToLS = (movies: Movie[]): void => {
  const favoriteMovies = localStorage.getItem(LSKey.FavoriteMovies);

  if (!favoriteMovies) {
    const preparedMovies = movies.reduce<FavoriteMovies>(
      (accumulator, movie) => {
        accumulator[movie.title] = false;

        return accumulator;
      },
      {}
    );

    localStorage.setItem(LSKey.FavoriteMovies, JSON.stringify(preparedMovies));
  }
};

export const getFavoriteMoviesConfigFromLS = (): FavoriteMovies => {
  const favoriteMovies = localStorage.getItem(LSKey.FavoriteMovies);

  return favoriteMovies ? JSON.parse(favoriteMovies) : {};
};

export const setFavoriteMoviesConfigToLS = (config: FavoriteMovies): void => {
  localStorage.setItem(LSKey.FavoriteMovies, JSON.stringify(config));
};

export const updateFavoriteMoviesConfig = (
  title: string
): FavoriteMovies => {
  const favoriteMovies = getFavoriteMoviesConfigFromLS();
  const newFavoriteMovies = {
    ...favoriteMovies,
    [title]: !favoriteMovies[title],
  };

  return newFavoriteMovies;
};
