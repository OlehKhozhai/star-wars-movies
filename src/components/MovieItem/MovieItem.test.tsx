import { create } from "react-test-renderer";

import { movieMock } from "utils/mock";
import MovieItem from "./MovieItem";

describe("MovieItem component", () => {
  it("should render by default", () => {
    const component = create(
      <MovieItem
        {...movieMock}
        isActive={false}
        isFavorite={false}
        onMovieClick={() => undefined}
      />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should render active movie", () => {
    const component = create(
      <MovieItem
        {...movieMock}
        isActive={true}
        isFavorite={false}
        onMovieClick={() => undefined}
      />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should render favorite and active movie", () => {
    const component = create(
      <MovieItem
        {...movieMock}
        isActive={true}
        isFavorite={true}
        onMovieClick={() => undefined}
      />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should trigger onMovieClick function", () => {
    const onMovieClickStub = jest.fn();
    const component = create(
      <MovieItem
        {...movieMock}
        isActive={true}
        isFavorite={true}
        onMovieClick={onMovieClickStub}
      />
    );

    component.root.findByType("li").props.onClick();

    expect(onMovieClickStub).toHaveBeenCalledWith(movieMock);
  });
});
