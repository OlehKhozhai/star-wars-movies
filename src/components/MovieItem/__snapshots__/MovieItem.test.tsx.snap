// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieItem component should render active movie 1`] = `
<li
  className="root active"
  onClick={[Function]}
>
  A New Hope
  <span
    className="starIcon"
  >
    <svg
      className=""
      fill="grey"
      height="25"
      width="23"
    >
      <polygon
        points="9.9, 1.1, 3.3, 21.78, 19.8, 8.58, 0, 8.58, 16.5, 21.78"
        style={
          Object {
            "fillRule": "nonzero",
          }
        }
      />
    </svg>
  </span>
</li>
`;

exports[`MovieItem component should render by default 1`] = `
<li
  className="root "
  onClick={[Function]}
>
  A New Hope
  <span
    className="starIcon"
  >
    <svg
      className=""
      fill="grey"
      height="25"
      width="23"
    >
      <polygon
        points="9.9, 1.1, 3.3, 21.78, 19.8, 8.58, 0, 8.58, 16.5, 21.78"
        style={
          Object {
            "fillRule": "nonzero",
          }
        }
      />
    </svg>
  </span>
</li>
`;

exports[`MovieItem component should render favorite and active movie 1`] = `
<li
  className="root active"
  onClick={[Function]}
>
  A New Hope
  <span
    className="starIcon"
  >
    <svg
      className=""
      fill="yellow"
      height="25"
      width="23"
    >
      <polygon
        points="9.9, 1.1, 3.3, 21.78, 19.8, 8.58, 0, 8.58, 16.5, 21.78"
        style={
          Object {
            "fillRule": "nonzero",
          }
        }
      />
    </svg>
  </span>
</li>
`;
